File: examples/python/phi4-mm.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License

import argparse
import os
import glob
import time
import json
from pathlib import Path

import onnxruntime_genai as og

# og.set_log_options(enabled=True, model_input_values=True, model_output_values=True)

def _find_dir_contains_sub_dir(current_dir: Path, target_dir_name):
    curr_path = Path(current_dir).absolute()
    target_dir = glob.glob(target_dir_name, root_dir=curr_path)
    if target_dir:
        return Path(curr_path / target_dir[0]).absolute()
    else:
        if curr_path.parent == curr_path:
            # Root dir
            return None
        return _find_dir_contains_sub_dir(curr_path / '..', target_dir_name)


def _complete(text, state):
    return (glob.glob(text + "*") + [None])[state]


def get_paths(modality, user_provided_paths, default_paths, interactive):
    paths = None

    if interactive:
        try:
            import readline
            readline.set_completer_delims(" \t\n;")
            readline.parse_and_bind("tab: complete")
            readline.set_completer(_complete)
        except ImportError:
            # Not available on some platforms. Ignore it.
            pass
        paths = [
            path.strip()
            for path in input(
                f"{modality.capitalize()} Path (comma separated; leave empty if no {modality}): "
            ).split(",")
        ]
    else:
        paths = user_provided_paths if user_provided_paths else default_paths

    paths = [path for path in paths if path]
    return paths


def run(args: argparse.Namespace):
    print("Loading model...")
    config = og.Config(args.model_path)
    if args.execution_provider != "follow_config":
        config.clear_providers()
        if args.execution_provider != "cpu":
            print(f"Setting model to {args.execution_provider}...")
            config.append_provider(args.execution_provider)
    model = og.Model(config)
    print("Model loaded")

    tokenizer = og.Tokenizer(model)
    processor = model.create_multimodal_processor()
    stream = processor.create_stream()

    interactive = not args.non_interactive

    while True:
        image_paths = get_paths(
            modality="image",
            user_provided_paths=args.image_paths,
            default_paths=[str(_find_dir_contains_sub_dir(Path(__file__).parent, "test") / "test_models" / "images" / "australia.jpg")],
            interactive=interactive
        )
        audio_paths = get_paths(
            modality="audio",
            user_provided_paths=args.audio_paths,
            default_paths=[str(_find_dir_contains_sub_dir(Path(__file__).parent, "test") / "test_models" / "audios" / "1272-141231-0002.mp3")],
            interactive=interactive
        )

        images = None
        audios = None

        # Validate and open image paths
        if len(image_paths) == 0:
            print("No image provided")
        else:
            for image_path in image_paths:
                if not os.path.exists(image_path):
                    raise FileNotFoundError(f"Image file not found: {image_path}")
                print(f"Using image: {image_path}")
            images = og.Images.open(*image_paths)

        # Validate and open audio paths
        if len(audio_paths) == 0:
            print("No audio provided")
        else:
            for audio_path in audio_paths:
                if not os.path.exists(audio_path):
                    raise FileNotFoundError(f"Audio file not found: {audio_path}")
                print(f"Using audio: {audio_path}")
            audios = og.Audios.open(*audio_paths)

        # Get prompt text
        if interactive:
            text = input("Prompt: ")
        else:
            text = args.prompt or "Does the audio summarize what is shown in the image? If not, what is different?"

        # Build multimodal content list
        content_list = []
        content_list.extend([{"type": "image"} for _ in image_paths])
        content_list.extend([{"type": "audio"} for _ in audio_paths])
        content_list.append({"type": "text", "text": text})

        # Construct messages and apply template
        messages = [{"role": "user", "content": content_list}]
        message_json = json.dumps(messages)
        prompt = tokenizer.apply_chat_template(message_json, add_generation_prompt=True)

        print("Processing inputs...")
        inputs = processor(prompt, images=images, audios=audios)
        print("Processor complete.")

        print("Generating response...")
        params = og.GeneratorParams(model)
        params.set_search_options(max_length=7680)

        generator = og.Generator(model, params)
        generator.set_inputs(inputs)
        start_time = time.time()

        while not generator.is_done():
            generator.generate_next_token()

            new_token = generator.get_next_tokens()[0]
            print(stream.decode(new_token), end="", flush=True)

        print()
        total_run_time = time.time() - start_time
        print(f"Total Time : {total_run_time:.2f}")

        for _ in range(3):
            print()

        # Delete the generator to free the captured graph before creating another one
        del generator

        if not interactive:
            break


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m", "--model_path", type=str, required=True, help="Path to the folder containing the model"
    )
    parser.add_argument(
        "-e", "--execution_provider", type=str, required=False, default='follow_config', choices=["cpu", "cuda", "dml", "follow_config"], help="Execution provider to run the ONNX Runtime session with. Defaults to follow_config that uses the execution provider listed in the genai_config.json instead."
    )
    parser.add_argument(
        "--image_paths", nargs='*', type=str, required=False, help="Path to the images, mainly for CI usage"
    )
    parser.add_argument(
        "--audio_paths", nargs='*', type=str, required=False, help="Path to the audios, mainly for CI usage"
    )
    parser.add_argument(
        '-pr', '--prompt', required=False, help='Input prompts to generate tokens from, mainly for CI usage'
    )
    parser.add_argument(
        '--non-interactive', action=argparse.BooleanOptionalAction, required=False, help='Non-interactive mode, mainly for CI usage'
    )
    args = parser.parse_args()
    run(args)

File: examples/c/src/phi4-mm.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <iostream>
#include <string>
#include <fstream>
#include <memory>
#include "common.h"
#include "ort_genai.h"

// C++ API Example

void CXX_API(
  GeneratorParamsArgs& generator_params_args,
  const std::string& model_path,
  const std::string& ep,
  const std::string& system_prompt,
  bool verbose,
  bool interactive
) {
  if (verbose) std::cout << "Creating config..." << std::endl;
  std::unordered_map<std::string, std::string> ep_options;
  auto config = GetConfig(model_path, ep, ep_options, generator_params_args);

  if (verbose) std::cout << "Creating model..." << std::endl;
  auto model = OgaModel::Create(*config);

  if (verbose) std::cout << "Creating multimodal processor..." << std::endl;
  auto processor = OgaMultiModalProcessor::Create(*model);

  if (verbose) std::cout << "Creating tokenizer..." << std::endl;
  auto tokenizer = OgaTokenizer::Create(*model);
  auto stream = OgaTokenizerStream::Create(*processor);

  while (true) {
    // Get images
    std::string image_paths_str;
    std::cout << "Image Path (comma separated; leave empty if no image):" << std::endl;
    std::getline(std::cin, image_paths_str);
    std::unique_ptr<OgaImages> images;
    std::vector<std::string> image_paths;
    for (size_t start = 0, end = 0; end < image_paths_str.size(); start = end + 1) {
      end = image_paths_str.find(',', start);
      image_paths.push_back(Trim(image_paths_str.substr(start, end - start)));
    }
    if (image_paths.empty()) {
      std::cout << "No image provided" << std::endl;
    } else {
      std::cout << "Loading images..." << std::endl;
      for (const auto& image_path : image_paths) {
        std::filesystem::path p(image_path);
        if (!std::filesystem::exists(p)) {
          throw std::runtime_error(std::string("Image file not found: ") + image_path);
        }
      }
      std::vector<const char*> image_paths_c;
      for (const auto& image_path : image_paths) image_paths_c.push_back(image_path.c_str());
      images = OgaImages::Load(image_paths_c);
    }

    // Get audios
    std::string audio_paths_str;
    std::cout << "Audio Path (comma separated; leave empty if no audio):" << std::endl;
    std::getline(std::cin, audio_paths_str);
    std::unique_ptr<OgaAudios> audios;
    std::vector<std::string> audio_paths;
    for (size_t start = 0, end = 0; end < audio_paths_str.size(); start = end + 1) {
      end = audio_paths_str.find(',', start);
      audio_paths.push_back(Trim(audio_paths_str.substr(start, end - start)));
    }
    if (audio_paths.empty()) {
      std::cout << "No audio provided" << std::endl;
    } else {
      std::cout << "Loading audios..." << std::endl;
      for (const auto& audio_path : audio_paths) {
        std::filesystem::path p(audio_path);
        if (!std::filesystem::exists(p)) {
          throw std::runtime_error(std::string("Audio file not found: ") + audio_path);
        }
      }
      std::vector<const char*> audio_paths_c;
      for (const auto& audio_path : audio_paths) audio_paths_c.push_back(audio_path.c_str());
      audios = OgaAudios::Load(audio_paths_c);
    }

    std::string text;
    std::cout << "Prompt: " << std::endl;
    std::getline(std::cin, text);

    // Construct messages string with special tokens for ApplyChatTemplate
    std::string content;
    for (size_t i = 0; i < image_paths.size(); ++i)
      content += "<|image_" + std::to_string(i + 1) + "|>\\n";
    for (size_t i = 0; i < audio_paths.size(); ++i)
      content += "<|audio_" + std::to_string(i + 1) + "|>\\n";
    content += text;

    const std::string messages = R"([{"role": "user", "content": ")" + content + R"("}])";

    std::string prompt = std::string(tokenizer->ApplyChatTemplate("", messages.c_str(), "", true));

    std::cout << "Processing images, audios, and prompt..." << std::endl;
    auto input_tensors = processor->ProcessImagesAndAudios(prompt.c_str(), images.get(), audios.get());

    std::cout << "Generating response..." << std::endl;
    auto params = OgaGeneratorParams::Create(*model);
    params->SetSearchOption("max_length", 7680);

    auto generator = OgaGenerator::Create(*model, *params);
    generator->SetInputs(*input_tensors);

    while (true) {
      generator->GenerateNextToken();

      if (generator->IsDone()) {
        break;
      }

      const auto num_tokens = generator->GetSequenceCount(0);
      const auto new_token = generator->GetSequenceData(0)[num_tokens - 1];
      std::cout << stream->Decode(new_token) << std::flush;
    }

    std::cout << "\n\n" << std::endl;
  }
}

int main(int argc, char** argv) {
  // Get command-line args
  GeneratorParamsArgs generator_params_args;
  GuidanceArgs guidance_args;
  std::string model_path, ep = "follow_config", system_prompt = "You are a helpful AI assistant.";
  bool verbose = false, interactive = true, rewind = false;
  if (!ParseArgs(argc, argv, generator_params_args, guidance_args, model_path, ep, system_prompt, verbose, interactive, rewind)) {
    return -1;
  }

  // Responsible for cleaning up the library during shutdown
  OgaHandle handle;

  std::cout << "--------------------" << std::endl;
  std::cout << "Hello, Phi-4-Multimodal!" << std::endl;
  std::cout << "--------------------" << std::endl;
  
  std::cout << "Model path: " << model_path << std::endl;
  std::cout << "Execution provider: " << ep << std::endl;
  std::cout << "System prompt: " << system_prompt << std::endl;
  std::cout << "Verbose: " << verbose << std::endl;
  std::cout << "Interactive: " << interactive << std::endl;
  std::cout << "--------------------------" << std::endl;
  std::cout << std::endl;

  try {
    CXX_API(generator_params_args, model_path, ep, system_prompt, verbose, interactive);
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return -1;
  }

  return 0;
}